controller.go

```diff
@@ -0,0 +1,149 @@
+	return &RandomSecretController{
+		context:               context,
+		randomSecretClientset: randomSecretClientset,
+	}
+}
+
+// Watch watches for instances of RandomSecret custom resources and acts on them
+func (c *RandomSecretController) StartWatch(namespace string, stopCh chan struct{}) error {
+
+	resourceHandlers := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+	restClient := c.randomSecretClientset.RESTClient()
+	watcher := opkit.NewWatcher(randomsecrets.RandomSecretResource, namespace, resourceHandlers, restClient)
+	go watcher.Watch(&randomsecrets.RandomSecret{}, stopCh)
+	return nil
+}
+
+func (c *RandomSecretController) generateData() []byte {
+	src := make([]byte, 64)
+	_, err := rand.Read(src)
+	if err != nil {
+		panic(err)
+	}
+	val := make([]byte, hex.EncodedLen(len(src)))
+	hex.Encode(val, src)
+	return val
+}
+
+func (c *RandomSecretController) updateSecret(rs *randomsecrets.RandomSecret, s *v1.Secret) {
+	blockOwnerDeletion := true
+	controller := true
+	s.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion:         randomsecrets.RandomSecretResource.Version,
+			BlockOwnerDeletion: &blockOwnerDeletion,
+			Controller:         &controller,
+			Kind:               randomsecrets.RandomSecretResource.Kind,
+			Name:               rs.Name,
+			UID:                rs.UID,
+		},
+	}
+	if s.Data == nil {
+		s.Data = make(map[string][]byte)
+	}
+	if val, ok := s.Data["random"]; !ok || len(val) < 128 {
+		fmt.Printf("Adding data to Secret '%s'...\n", s.Name)
+		s.Data["random"] = c.generateData()
+	}
+}
+
+func (c *RandomSecretController) ensureSecret(obj *randomsecrets.RandomSecret) {
+	secretsClient := c.context.Clientset.CoreV1().Secrets(obj.Namespace)
+	s, err := secretsClient.Get(obj.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		fmt.Printf("Creating Secret '%s'...\n", obj.Name)
+		s = &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      obj.Name,
+				Namespace: obj.Namespace,
+			},
+		}
+		c.updateSecret(obj, s)
+		_, err := secretsClient.Create(s)
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			s, getErr := secretsClient.Get(obj.Name, metav1.GetOptions{})
+			if getErr != nil {
+				panic(fmt.Errorf("Failed to get Secret '%s': %v", s.Name, getErr))
+			}
+
+			c.updateSecret(obj, s)
+			_, updateErr := secretsClient.Update(s)
+			return updateErr
+		})
+		if retryErr != nil {
+			panic(fmt.Errorf("Update failed: %v", retryErr))
+		}
+	}
+}
+
+func (c *RandomSecretController) onAdd(obj interface{}) {
+	s := obj.(*randomsecrets.RandomSecret).DeepCopy()
+	fmt.Printf("Add event: RandomSecret '%s/%s' with Spec=%s...\n", s.Namespace, s.Name, s.Spec)
+
+	c.ensureSecret(s)
+}
+
+func (c *RandomSecretController) onUpdate(oldObj, newObj interface{}) {
```

Comment:
seems like duplication here...is there not some way we can have both methods be the same?

Reply:
I'm not sure, really. I felt this was a fairly minimal amount of duplication